src/pytest_split/plugin.py

```diff
@@ -76,130 +79,166 @@ def pytest_cmdline_main(config: "Config") -> None:
-    def __init__(self):
-        self._suite: TestSuite
-        self._group: TestGroup
-        self._messages: "List[str]" = []
 
-    def pytest_report_collectionfinish(self, config: "Config") -> "List[str]":
-        lines = []
-        if self._messages:
-            lines += self._messages
+def pytest_configure(config: "Config") -> None:
+    """
+    Enable the plugins we need.
+    """
+    if config.option.splits and config.option.group:
+        config.pluginmanager.register(PytestSplitPlugin(config), "pytestsplitplugin")
 
-        if hasattr(self, "_suite"):
-            lines.append(
-                f"Running group {self._group.index}/{self._suite.splits}"
-                f" ({self._group.num_tests}/{self._suite.num_tests}) tests"
-            )
+    if config.option.store_durations:
+        config.pluginmanager.register(PytestSplitCachePlugin(config), "pytestsplitcacheplugin")
 
-        prefix = "[pytest-split]"
-        lines = [f"{prefix} {m}" for m in lines]
-
-        return lines
-
-    def pytest_collection_modifyitems(
-        self, config: "Config", items: "List[nodes.Item]"
-    ) -> None:
-        splits = config.option.splits
-        group = config.option.group
-        store_durations = config.option.store_durations
-        durations_report_path = config.option.durations_path
-
-        if store_durations:
-            if any((group, splits)):
-                self._messages.append(
-                    "Not splitting tests because we are storing durations"
-                )
-            return None
-
-        if not group and not splits:
-            # don't split unless explicitly requested
-            return None
-
-        if not os.path.isfile(durations_report_path):
-            self._messages.append(
-                "Not splitting tests because the durations_report is missing"
-            )
-            return None
 
-        with open(durations_report_path) as f:
-            stored_durations = OrderedDict(json.load(f))
+class Base:
+    def __init__(self, config: "Config") -> None:
+        """
+        Load durations and set up a terminal writer.
 
-        start_idx, end_idx = _calculate_suite_start_and_end_idx(
-            splits, group, items, stored_durations
-        )
+        This logic is shared for both the split- and cache plugin.
+        """
+        self.config = config
+        self.writer = create_terminal_writer(self.config)
 
-        self._suite = TestSuite(splits, len(items))
-        self._group = TestGroup(group, end_idx - start_idx)
-        items[:] = items[start_idx:end_idx]
+        try:
+            with open(config.option.durations_path, "r") as f:
+                self.cached_durations = json.loads(f.read())
+        except FileNotFoundError:
+            self.cached_durations = {}
 
+        # This code provides backwards compatibility after we switched
+        # from saving durations in a list-of-lists to a dict format
+        # Remove this when bumping to v1
```

Comment:
💪 

Is there a test case for this?

Reply:
I'll add one 👍

Reply:
Should probably add a 100% coverage requirement as well at some point 🤔

Reply:
Do it tomorrow 😂 

Added the old format to test in 794fedf

Reply:
I could help you write some tests if you want, and if you'd like I could also help implement [poetry](https://python-poetry.org/) for package management and dependency stuff 🙂

Reply:
Poetry would be welcome indeed 👍